Documentation reorganisation (#183)
esl/MongoosePush: guides/docker.md

# Docker

## Running from DockerHub

We provide prebuilt MongoosePush images. Configuration requires either an FCM token, APNS certificates or an APNS token. Depending on your usecase, you can have some or all of them in a standalone MongoosePush instance or using a docker container.

> For the full configuration you need to set the following directory structure up:

Reviewer: ```suggestion
For the full configuration you need to set up the following directory structure:
```

> For the full configuration you need to set up the following directory structure:

Reviewer 2: I knew I've seen that before:
https://github.com/esl/MongoosePush/pull/103#discussion_r322683529
```suggestion
For the full configuration you need to set the following directory structure up:
```

Author: And I saw it in this PR already [HERE](https://github.com/esl/MongoosePush/pull/183#discussion_r448946353)! 🤷‍♀️

Reviewer: "set up" is correct here. @kmakiela, you originally had "setup", which is a noun, not a verb ("set up" vs "setup"), in the position of a verb. @goddammit corrected that to a "set something up" construction, which is correct, but in this particular example I find that inner "something" too long and pushes the two particles of the verb too far away from each other, so I thought of "set up something" instead.

* priv/
    * ssl/
      * rest_cert.pem - The HTTP endpoint certificate
      * rest_key.pem - private key for the HTTP endpoint certificate (has to be unencrypted)
    * apns/
      * prod_cert.pem - Production APNS app certificate
      * prod_key.pem - Production APNS app certificate's private key (has to be unencrypted)
      * dev_cert.pem - Development APNS app certificate
      * dev_key.pem - Development APNS app certificate's private key (has to be unencrypted)
      * token.p8 - `APNS` authentication token
    * fcm/
      * token.json - `FCM` service account JSON file

> If you want to use `APNS` token authentication you need to provide token and set `key_id` and `team_id` environmental variables. To see how to obtain token and `key_id` read:

Reviewer 3: ```suggestion
If you want to use the `APNS` token authentication you need to provide the token and set both `key_id` and `team_id` environmental variables. To see how to obtain the token and `key_id` read:
```

> If you want to use the `APNS` token authentication you need to provide the token and set both `key_id` and `team_id` environmental variables. To see how to obtain the token and `key_id` read:
> https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_token_based_connection_to_apns

Reviewer 2: ```suggestion
If you want to use the `APNS` token authentication you need to provide the token and set both `key_id` and `team_id` environmental variables. To see how to obtain the token and `key_id` read [this](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_token_based_connection_to_apns)
```

> To see how to obtain `team_id` read: https://www.mobiloud.com/help/knowledge-base/ios-app-transfer/

Reviewer 2: ```suggestion
To see how to obtain `team_id` read [this](https://www.mobiloud.com/help/knowledge-base/ios-app-transfer/)
```

> `FCM` JSON file can be generated by Firebase console (https://console.firebase.google.com). Go to your project -> `Project Settings` -> `Service accounts` -> `Generate new private key`

Reviewer 3: ```suggestion
`FCM` JSON file can be generated by the Firebase console (https://console.firebase.google.com). Go to your project -> `Project Settings` -> `Service accounts` -> `Generate new private key`
```

> `FCM` JSON file can be generated by the Firebase console (https://console.firebase.google.com). Go to your project -> `Project Settings` -> `Service accounts` -> `Generate new private key`

Reviewer: ```suggestion
The `FCM` JSON file can be generated by the Firebase console (https://console.firebase.google.com). Go to your project -> `Project Settings` -> `Service accounts` -> `Generate new private key`.
```

> Assuming that you have the `priv` directory with all certificates and fcm token in current directory, then you may start MongoosePush with the following command:

Reviewer 3: ```suggestion
Assuming that you have the `priv` directory with all certificates and the fcm token in the current directory, you may start MongoosePush with the following command:
```

```bash
docker run -v `pwd`/priv:/opt/app/priv \
  -e PUSH_HTTPS_CERTFILE="/opt/app/priv/ssl/rest_cert.pem" \
  -e PUSH_HTTPS_KEYFILE="/opt/app/priv/ssl/rest_key.pem" \
  -it --rm mongooseim/mongoose-push:latest
```

## Building

> Building docker is really easy, just type:

Reviewer: ```suggestion
Building a docker image is really easy, just type:
```

```bash
docker build . -t mpush:latest
```

> As a result of this command you get access to `mpush:latest` docker image. You may run it by typing:

Reviewer: ```suggestion
As a result of this command you get access to the `mpush:latest` docker image. You may run it by typing:
```

```bash
docker run -it --rm mpush:latest foreground
```

The docker image that you have just built, exposes the port `8443` for the HTTP API of MongoosePush. It contains a `VOLUME` for path */opt/app* - it is handy for injecting `APNS` and `HTTP API` certificates since by default the docker image comes with test, self-signed certificates.

## Configuration (basic)

> The docker image of MongoosePush contains common, basic configuration that is generated from `config/prod.exs`. All useful options may be overridden via system environmental variables. Below there's a full list of the variables you may set while running docker (via `docker -e` switch), but if there's something you feel, you need to change other then that, then you need to prepare your own `config/prod.exs` before image build.

Reviewer: ```suggestion
The docker image of MongoosePush contains common, basic configuration that is generated from `config/prod.exs`. All useful options may be overridden via environment variables. Below there's a full list of the variables you may set while running docker (via `docker -e`), but if there's something you feel you need to change other then that, then you need to prepare your own `config/prod.exs` before image build.
```

> The docker image of MongoosePush contains common, basic configuration that is generated from `config/prod.exs`. All useful options may be overridden via system environmental variables. Below there's a full list of the variables you may set while running docker (via `docker -e` switch), but if there's something you feel, you need to change other then that, then you need to prepare your own `config/prod.exs` before image build.

Reviewer: Also, it says "Below there's a full list of...", but this is the end of the article, where is that full list? 😅

> The docker image of MongoosePush contains common, basic configuration that is generated from `config/prod.exs`. All useful options may be overridden via environment variables. Below there's a full list of the variables you may set while running docker (via `docker -e`), but if there's something you feel you need to change other then that, then you need to prepare your own `config/prod.exs` before image build.

Reviewer 2: There is no "Below" anymore. Maybe create a ref to the config guide?

Author: Done in the meantime:)
